[css-fonts-4] Contribute Microsoft tests for Font Variations

FremyCompany opened the pull request:
Contributing (most) of our own CSS Fonts 4 tests to web-platform-tests. Some tests have not been ported because they require Microsoft Windows fonts and I have not enough time/expertise to create a dummy version of these fonts.

**INTEROP:** Chrome, Safari and Firefox all seem to have issues with these tests, so it would be good if someone from either/all of these browsers could review the tests. 

**NOTE:** One of the test requires you to install some fonts natively on your system. These fonts are a new (but backward-compatible) version of the fonts that are already required for other css-fonts tests. I created this new version because the old version didn't give any mean to test font matching via JavaScript, which is annoying. If this PR gets merged, someone should update http://www.w3.org/Style/CSS/Test/Fonts/CSSTest/ but I don't know who to contact.

<!-- Reviewable:start -->

<!-- Reviewable:end -->

FremyCompany committed:
[css-fonts-4] Contribute Microsoft tests for Font Variations

w3c-bots commented:
# Build PENDING

Started: None
Finished: None


<h2>Failing Jobs</h2>
<ul>

<li>chrome:dev</li>

</ul>




View more information about this build on:

- [WPT PR Status](http://pulls.web-platform-tests.org/build/23822)
- [TravisCI](https://travis-ci.org/w3c/web-platform-tests/builds/336770565)

SergeyMalkin commented on css/css-fonts/variations/font-opentype-collections.html:
```diff
@@ -0,0 +1,38 @@
+<!DOCTYPE html>
```
This is not variable fonts test, so you may want to move it outside variations folder. But otherwise test is useful and should also be expanded to test access to individual fonts inside collection.

SergeyMalkin commented on css/css-fonts/variations/font-shorthand.html:
```diff
@@ -0,0 +1,66 @@
+
+    <script>
+        testFontShorthand = [
+            { value: "calc(24px) Arial",                                     isValid:true,   message: "Font weight specified as number" },
```
Comment is incorrect. This should be something like "font-weight is not part of shorthand"

SergeyMalkin commented on css/css-fonts/variations/font-style.html:
```diff
@@ -0,0 +1,137 @@
+
+        });
+
+        async_test(function (test) {
```
This test is already in separate file. You can remove it from here.

SergeyMalkin replied:
Same for transition test below

SergeyMalkin commented on css/css-fonts/variations/font-weight-interpolation.html:
```diff
@@ -0,0 +1,84 @@
+    <script src="/resources/testharness.js"></script>
+    <script src="/resources/testharnessreport.js"></script>
+    <style>
+        @keyframes fontStyleAnimation {
```
fontStyleAnimation -> fontWeightAnimation

SergeyMalkin commented on css/css-fonts/variations/font-variation-settings-parsing.html:
```diff
@@ -0,0 +1,54 @@
+    <script src="/resources/testharnessreport.js"></script>
+</head>
+<body>
+    <script>
```
Add "use strict"

SergeyMalkin commented on css/css-fonts/variations/font-variation-settings-parsing.html:
```diff
@@ -0,0 +1,54 @@
+<body>
+    <script>
+        valueParserTests = [
+            { input: "'wght' 1000, '9 ~A' -45",           output: "\"9 ~A\" -45, \"wght\" 1000",    valid: true,  message: "Axis tag with valid non-letter ascii characters" },
```
Would be nice to have names consistent between tests, e.g. input/output/valid->value/expectedComputedValue/isValid

SergeyMalkin commented on css/css-fonts/variations/font-weight-lighter-bolder.html:
```diff
@@ -0,0 +1,57 @@
+		</div>
+    </div>
+	
+    <script>
```
"use strict"

SergeyMalkin commented on css/css-fonts/variations/font-weight-parsing.html:
```diff
@@ -0,0 +1,51 @@
+            { weight: "1000.001", isValid: false, message: "Values above maximum should be rejected" },
+            { weight: "calc(100.5)", isValid: true, expectedStyle: "100.5", message: "Simple calc value" },
+            { weight: "calc(1001)", isValid: false, message: "Out-of-range simple calc value" },
+            { weight: "calc(100.5*3 + 50.5)", isValid: true, expectedStyle: "352",  message: "Valid calc expression" },
```
expectedStyle -> expectedWeight

FremyCompany commented:
Linting errors I will have to fix:
- TRAILING WHITESPACE: 1 
- AHEM COPY: 1 
- CR AT EOL: 940 
- MISSING-LINK: 13

SergeyMalkin commented:
You forgot to port two files: extensive set of font matching tests and also tests for parsing @font-face descriptors.

emilio commented:
cc @jfkthame

FremyCompany committed:
Address Sergey's feedback (part 1)

FremyCompany committed:
Address Sergey's feedback (part 2)

SergeyMalkin commented on css/css-fonts/variations/font-style-parsing.html:
```diff
@@ -33,7 +33,10 @@
-            { style: "oblique -a",        expectedResult: false, message: "'oblique' followed by minus and non-number is invalid" }
+            { style: "oblique -a",        expectedResult: false, message: "'oblique' followed by minus and non-number is invalid" },
+            { style: "oblique calc(50deg)",     expectedResult: true,  message: "'oblique' followed by calc is valid" },
+            { style: "oblique calc(-120deg)",   expectedResult: true,  message: "'oblique' followed by calc is valid even if it must be clamped (no computation)" },
```
Can you add cases for different angle units? E.g. calc(2 * 0.1rad)

FremyCompany replied:
I doubt that this is relevant. A browser that would fail these tests but not the existing ones would have an issue with css-value-and-units not font-style parsing.

SergeyMalkin replied:
Are there tests for css-values to cover calc() for angles? I don't see any, but I may be missing it. 

There are specific tests covering other properties and units. We should have tests for angles, either here or in css-values folder

SergeyMalkin commented on css/css-fonts/variations/font-style-parsing.html:
```diff
@@ -33,7 +33,10 @@
-            { style: "oblique -a",        expectedResult: false, message: "'oblique' followed by minus and non-number is invalid" }
+            { style: "oblique -a",        expectedResult: false, message: "'oblique' followed by minus and non-number is invalid" },
+            { style: "oblique calc(50deg)",     expectedResult: true,  message: "'oblique' followed by calc is valid" },
+            { style: "oblique calc(-120deg)",   expectedResult: true,  message: "'oblique' followed by calc is valid even if it must be clamped (no computation)" },
```
Are you calc(-120deg) is valid value? My understanding is that final calculated value is tested against acceptable range, otherwise it is treated as parse error.

FremyCompany replied:
https://drafts.csswg.org/css-values/#calc-range

SergeyMalkin replied:
Ah, ok. Then both Edge and Chrome have the same bug across font-weight/stretch/style.

Can you review font-weight and font-stretch as well?

FremyCompany committed:
More linting fixup

FremyCompany committed:
Added missing element

FremyCompany committed:
Fixed incorrect calc() tests -- values should be clamped not rejected

drott commented:
> INTEROP: Chrome, Safari and Firefox all seem to have issues with these tests, so it would be good if someone from either/all of these browsers could review the tests.

Thanks for upstreaming and contributing these, @FremyCompany. I will put it on my TODO list to go through them for Blink and leave review feedback accordingly, but it will take a bit more time.

FremyCompany commented:
@drott Thanks, totally understood

drott reviewed (changes requested):
An excellent quality set of test cases - very helpful, thank you! Running these in Blink already uncovered 9 issues (compare [this meta bug](https://bugs.chromium.org/p/chromium/issues/detail?id=809920)).

I got curious about these test cases so I gave them a go in Blink. 2 pass as is, 10 fail, some of them only with minor issus.

For the test cases that partially need system fonts, I would suggest to split those out into separate tests for making it easier to compare results across engines, as the manual installation / harness preparation with the system fonts is an extra step that automated testing might not cover well yet.

I don't agree with the expectations for `css/css-fonts/variations/font-variation-settings-inherit.html`. But this might be only my interpretation of the spec or I might have missed something. It seems we need a spec clarification here. If you'd like to merge this PR quickly, I would suggest to isolate this one into a separate PR until we work out with @litherum   and others what the expected behavior should be.

Some test cases trigger a harness error / warning for duplicate test names. I believe this is minor but probably worth fixing.

Thanks again, I am happy to straighten out the Blink issues to pass more of these new tests.

FremyCompany commented on css/css-fonts/variations/at-font-face-font-matching.html:
```diff
@@ -0,0 +1,135 @@
+    <link rel="help" href="https://www.w3.org/TR/css-fonts-4/#font-matching-algorithm" />
+    <script src="../../../../resources/testharness.js"></script>
+    <script src="../../../../resources/testharnessreport.js"></script>
+    <!-- PART OF THIS TEST REQUIRES THAT YOU INSTALL THE [csstest-*.ttf] FONTS OF THE [resources] FOLDER -->
```
Seems reasonable, will do

FremyCompany replied:
✔ (this test didn't require the fonts to be installed, I converted it to use linked fonts)

FremyCompany commented on css/css-fonts/variations/font-shorthand.html:
```diff
@@ -0,0 +1,69 @@
+            { value: "1700.5 24px Arial",                                   isValid:false,  message: "Font weight specified as number, value greater than 1000" },
+
+            // font-weight as calc()
+            { value: "calc(900.5 - 200.1 * 2) calc(12px + 12px) Arial",     isValid:true,   expectedWeight:"500.3", message: "Font weight specified as calc()" },
```
Yes, I can make the calc land on 0.5 multiple.

FremyCompany replied:
✔

FremyCompany commented on css/css-fonts/variations/font-shorthand.html:
```diff
@@ -0,0 +1,69 @@
+
+            // font-weight as calc()
+            { value: "calc(900.5 - 200.1 * 2) calc(12px + 12px) Arial",     isValid:true,   expectedWeight:"500.3", message: "Font weight specified as calc()" },
+            { value: "calc(400.5 - 200.1 * 2) 24px Arial",                  isValid:false,                          message: "Font weight specified as calc(), value less than 1" },
```
On second look though, this expression should be valid though, because of calc() clamping. 
I will update the expectation here ^_^

FremyCompany replied:
✔

FremyCompany commented on css/css-fonts/variations/font-shorthand.html:
```diff
@@ -0,0 +1,69 @@
+            // font-style
+            { value: "oblique 45deg 24px Arial",                            isValid:true,   expectedStyle: "oblique 45deg",  message: "'oblique' with positive angle" },
+            { value: "oblique -45deg 24px Arial",                           isValid:true,   expectedStyle: "oblique -45deg", message: "'oblique' with hegative angle" },
+            { value: "oblique 24px Arial",                                  isValid:true,   expectedStyle: "oblique",        message: "'oblique' without slant angle" },
```
I think it's fine to leave the test strict here, but it's also probably ok not to pass that specific test, I doubt this is going to cause much webcompat trouble ^_^

I think there's still supposed to be a different between oblique and italic if a font supports both. It seems `oblique` shoud (I *think* but I'm not entirely sure) prefer any oblique fontface over the italic fontface, while italic should prefer any italic version of an oblique fontface.

I have never seen a font with both an oblique and an italic variant though, so I am not sure how to test whether this is true in any browser. Maybe the LaTeX font?

FremyCompany replied:
✔ (no change)

jfkthame replied:
Yes, the Computer Modern (or the newer Latin Modern) family offers both oblique ("slanted") and italic, and indeed they are used contrastively by some authors (e.g. Knuth himself).

(It also has an Unslanted Italic face, which AFAICS is not something we could express with CSS according to the current draft. I suppose something like

    font-style: italic oblique(0deg);

would make sense, if we want to allow it, but currently that would be a parse error.)

CSS Fonts 3 made it quite clear that 'italic' and 'oblique' could be distinguished, at least in `@font-face` families even if the UA doesn't support a distinction for system font families. I don't see any reason why we should regress that in Fonts 4.

svgeesus replied:
Indeed, Fonts 4 makes it clear that `oblique` takes an optional second parameter, the oblique angle; and  `oblique` now expands to `oblique 20deg`.

FremyCompany commented on css/css-fonts/variations/font-variation-settings-inherit.html:
```diff
@@ -0,0 +1,41 @@
+
+        test(() => {
+            var actualValue = window.getComputedStyle(elementParent).fontVariationSettings;
+            assert_equals(actualValue, "\"yyyy\" 2", "Duplicate axis tags should be removed, favoring the latter axis tag's value.");
```
It might be a result of how we store the flags in Edge, maybe we just can't store this duplication internally and therefore simplify anyway. Paging @SergeyMalkin here.

csnardi replied:
Drive-by comment: I believe the spec should soon be updated to reflect this as correct behavior per https://github.com/w3c/csswg-drafts/issues/1959.

FremyCompany commented on css/css-fonts/variations/font-weight-parsing.html:
```diff
@@ -0,0 +1,52 @@
+
+        var testContinuousWeights = [
+            { weight: "401",      isValid: true,  message: "Values that are not multiple of 100 should be parsed successfully" },
+            { weight: "400.1",    isValid: true,  message: "Non-integer Values should be parsed successfully" },
```
I will change to 400.5

FremyCompany replied:
✔

FremyCompany commented on css/css-fonts/variations/font-weight-matching.html:
```diff
@@ -0,0 +1,162 @@
+    <script src="/resources/testharness.js"></script>
+    <script src="/resources/testharness.js"></script>
+    <script src="/resources/testharnessreport.js"></script>
+    <!-- PART OF THIS TEST REQUIRES THAT YOU INSTALL THE [csstest-*.ttf] FONTS OF THE [resources] FOLDER -->
```
Seems reasonable

FremyCompany commented on css/css-fonts/variations/font-shorthand.html:
```diff
@@ -0,0 +1,69 @@
+    <script>
+
+        testFontShorthand = [
+            { value: "calc(24px) Arial",                                     isValid:true,   message: "Font weight specified as number" },
```
✔

FremyCompany commented on css/css-fonts/variations/font-shorthand.html:
```diff
@@ -0,0 +1,69 @@
+            { value: "calc(400.5 + 300.1 * 2) 24px Arial",                  isValid:false,                          message: "Font weight specified as calc(), value greater than 1000" },
+
+            // font-style
+            { value: "oblique 45deg 24px Arial",                            isValid:true,   expectedStyle: "oblique 45deg",  message: "'oblique' with positive angle" },
```
✔ (no change)

FremyCompany commented on css/css-fonts/variations/font-stretch.html:
```diff
@@ -0,0 +1,86 @@
+    <script>
+
+        testStretchValues = [
+            { stretch: "100",               expectedComputedStretch: "100%" ,   expectedIsSupported: false, message: "only percentages, not numbers allowed" },
```
✔ (the same test was just duplicated)

FremyCompany commented on css/css-fonts/variations/font-style-parsing.html:
```diff
@@ -0,0 +1,49 @@
+            { style: "oblique - 20deg",   expectedResult: false, message: "'oblique' followed by minus and angle separated by space is invalid" },
+            { style: "oblique -a",        expectedResult: false, message: "'oblique' followed by minus and non-number is invalid" },
+            { style: "oblique calc(50deg)",     expectedResult: true,  message: "'oblique' followed by calc is valid" },
+            { style: "oblique calc(-120deg)",   expectedResult: true,  message: "'oblique' followed by calc is valid even if it must be clamped (no computation)" },
```
✔ (no change)

FremyCompany commented on css/css-fonts/variations/font-variation-settings-inherit.html:
```diff
@@ -0,0 +1,41 @@
+<!DOCTYPE html>
```
✔ (no change)

FremyCompany commented on css/css-fonts/variations/font-style-parsing.html:
```diff
@@ -0,0 +1,49 @@
+            { style: "oblique -a",        expectedResult: false, message: "'oblique' followed by minus and non-number is invalid" },
+            { style: "oblique calc(50deg)",     expectedResult: true,  message: "'oblique' followed by calc is valid" },
+            { style: "oblique calc(-120deg)",   expectedResult: true,  message: "'oblique' followed by calc is valid even if it must be clamped (no computation)" },
+            { style: "oblique calc(6 * 20deg)", expectedResult: true,  message: "'oblique' followed by calc is valid even if it must be clamped (with computation)" },
```
✔ (no change)

FremyCompany commented on css/css-fonts/variations/font-variation-settings-parsing.html:
```diff
@@ -0,0 +1,57 @@
+    <script>
+
+        var valueParserTests = [
+            { value: "'wght' 1000, '9 ~A' -45",           expectedComputedValue: "\"9 ~A\" -45, \"wght\" 1000",    isValid: true,  message: "Axis tag with valid non-letter ascii characters" },
```
✔

csnardi replied:
This still seems to exist, looking at the Travis log:
```
`Property value: Percentages should not be supported` | **FAIL: 20/10** 
`@supports: Percentages should not be supported` | **PASS: 20/10**
```

FremyCompany committed:
Addressed @drott's feedback

FremyCompany commented:
> For the test cases that partially need system fonts, I would suggest to split those out into separate tests for making it easier to compare results across engines, as the manual installation / harness preparation with the system fonts is an extra step that automated testing might not cover well yet.

Done. 

I also converted to url() fonts some tests that unnecessarily relied on the fonts being available. 

> I don't agree with the expectations for css/css-fonts/variations/font-variation-settings-inherit.html. But this might be only my interpretation of the spec or I might have missed something. It seems we need a spec clarification here. If you'd like to merge this PR quickly, I would suggest to isolate this one into a separate PR until we work out with @litherum and others what the expected behavior should be.

I updated the test to have a generic-enough assertion it passes in both browsers. We can always refine in a further pull request when we have the expected behavior clarified in the spec.

drott reviewed (approved):
Thanks for addressing the comments, LGTM. I'll file an issue for the inheritance / precedence of duplicate axis tags questions.

drott commented on css/css-fonts/variations/font-variation-settings-inherit.html:
```diff
@@ -0,0 +1,48 @@
+
+        test(() => {
+            var actualValue = window.getComputedStyle(elementParent).fontVariationSettings;
+            // The following strict test is subject to debate; softening for now:
```
Filed https://github.com/w3c/csswg-drafts/issues/2288

svgeesus replied:
@drott did you want to hold off on approving all tests while that issue gets discussed, or approve all but that test, or approve all and then fix up if needed?

drott replied:
I am good with merging the more lenient version of this test and discussing on the separate issue in the meantime. 

I did approve the PR, see above "drott approved these changes 23 hours ago", but it seems like that doesn't count 😢.

"Review required
At least one approved review is required by reviewers with write access. "

Would I be eligible for write access, who could configure it for me?

svgeesus replied:
> Would I be eligible for write access, who could configure it for me?

I just did, https://github.com/w3c/web-platform-tests/pull/9525 but now that tiny edit needs a reviewer ...

svgeesus commented:
@FremyCompany 
> If this PR gets merged, someone should update http://www.w3.org/Style/CSS/Test/Fonts/CSSTest/ but I don't know who to contact.

That would be me. But discuss with CSS WG first as there have been several updates to Ahem over recent years and we want to be sure the new one is a true superset of the most recent previous update (plus there was quite some discussion as I recall).

FremyCompany commented:
@svgeesus I didn't update Ahem so that shouldn't be a problem. 

I updated [the CSSTest fonts that are currently provided at the url pointed above](http://www.w3.org/Style/CSS/Test/Fonts/CSSTest/). I manually checked all the tests that are using those fonts, and they didn't regress based on the update, so I think we should be fine.

svgeesus reviewed (changes requested):
Approved (once the build passes)

FremyCompany committed:
Fixed lint issues

FremyCompany committed:
Remove fallback to linked font in installed-font test cases

FremyCompany commented:
@svgeesus The build now passes everything but the Firefox run. It does not fail because of me, it fails because it cannot seem to start the test. After 24mins, it gave up. I think we should merge this, and someone from the infra team should take a look at what is going on.

I can't merge because your review is marked as "requiring changes" though.

drott commented:
@svgeesus ping? Can you help getting this merged, thank you!

csnardi commented on css/css-fonts/variations/font-style-parsing.html:
```diff
@@ -0,0 +1,49 @@
+            { style: "oblique 90deg",     expectedResult: true,  message: "'oblique' followed by maxumum 90 degree angle is valid" },
+            { style: "oblique -90deg",    expectedResult: true,  message: "'oblique' followed by minimum -90 degree angle is valid" },
+            { style: "oblique 90.01deg",  expectedResult: false, message: "'oblique' followed by positive out of range angle is in invalid" },
+            { style: "oblique -90.01deg", expectedResult: false, message: "'oblique' followed by positive out of range angle is in invalid" },
```
Another duplicate test name:

```
`Font-style: 'oblique' followed by positive out of range angle is in invalid` | **PASS: 20/10**
```

FremyCompany replied:
Ah, that's what 20/10 means. That wasn't clear to me. Hopefully fixed now?

csnardi reviewed (commented):
It looks like the Firefox build was actually failing, which would prevent this from being merged.

FremyCompany committed:
Fixed some duplicate test names

FremyCompany commented:
Thanks @csnardi that was indeed it 🙌
@svgeesus would you mind merging the pull request now?

FremyCompany commented:
Up @svgeesus
